easybuild/easyblocks/o/ocaml.py

```diff
@@ -0,0 +1,129 @@
+##
+# Copyright 2015-2015 Ghent University
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://vscentrum.be/nl/en),
+# the Hercules foundation (http://www.herculesstichting.be/in_English)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# http://github.com/hpcugent/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for building and installing OCaml + opam (+ extensions), implemented as an easyblock
+
+@author: Kenneth Hoste (Ghent University)
+"""
+import os
+
+import easybuild.tools.environment as env
+from easybuild.easyblocks.generic.configuremake import ConfigureMake
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.run import run_cmd
+
+
+EXTS_FILTER_OCAML_PACKAGES = ("eval `opam config env` && opam list --installed %(ext_name)s.%(ext_version)s", '')
+OPAM_SUBDIR = 'opam'
+
+
+class EB_OCaml(ConfigureMake):
+    """Support for building/installing OCaml + opam (+ additional extensions)."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialisation of custom class variables for OCaml."""
+        super(EB_OCaml, self).__init__(*args, **kwargs)
+        self.with_opam = False
+
+    def configure_step(self):
+        """Custom configuration procedure for OCaml."""
+        self.cfg['prefix_opt'] = '-prefix '
+        self.cfg.update('configopts', '-cc "%s %s"' % (os.environ['CC'], os.environ['CFLAGS']))
+
+        if not 'world.opt' in self.cfg['buildopts']:
+            self.cfg.update('buildopts', 'world.opt')
+
+        super(EB_OCaml, self).configure_step()
+
+    def install_step(self):
+        """
+        Custom install procedure for OCaml.
+        First install OCaml using 'make install', then install OPAM (if sources are provided).
```

Comment:
yes, there's pretty much no other way to install OCaml packages (somewhat reliably)